src/core/tsi/s2a/handshaker/s2a_tsi_handshaker.cc

```diff
@@ -16,34 +16,327 @@
  *
  */
 
+#include <grpc/support/port_platform.h>
+
 #include "src/core/tsi/s2a/handshaker/s2a_tsi_handshaker.h"
 
-namespace grpc {
+#include <grpc/support/alloc.h>
+#include <grpc/support/log.h>
+#include <grpc/support/string_util.h>
+#include <grpc/support/sync.h>
+
+#include "src/core/lib/slice/slice_internal.h"
+#include "src/core/lib/surface/channel.h"
+#include "src/core/tsi/s2a/handshaker/s2a_handshaker_client.h"
+#include "src/core/tsi/s2a/s2a_constants.h"
+
+namespace grpc_core {
 namespace experimental {
 
+/** The main struct for the S2A TSI handshaker. **/
+struct s2a_tsi_handshaker {
+  tsi_handshaker base;
+  grpc_slice target_name;
+  bool is_client;
+  bool has_sent_start_message;
+  bool has_created_handshaker_client;
+  char* handshaker_service_url;
+  grpc_pollset_set* interested_parties;
+  grpc_s2a_credentials_options* options;
+  s2a_handshaker_client_vtable* client_vtable_for_testing;
+  grpc_channel* channel;
+  gpr_mu mu;
+  s2a_handshaker_client* client;
+  bool shutdown;
+};
+
+/** The main struct for the S2A TSI handshaker result. **/
+typedef struct s2a_tsi_handshaker_result {
+  tsi_handshaker_result base;
+  uint16_t tls_version;
+  uint16_t tls_ciphersuite;
+  uint8_t* in_traffic_secret;
+  uint8_t* out_traffic_secret;
+  size_t traffic_secret_size;
+  char* spiffe_id;
+  char* hostname;
```

Comment:
Do we still need `hostname` now that we have `spiffe_id` to represent the identity?

Reply:
In the proto, we have left the option to use either `hostname` or `spiffe_id`, and so there is the same option here.

Reply:
I see. Can we add some comments about `or`?

Reply:
Sure - I can add that we use whatever field is populated by the S2A service in the peer identity message. 

Do you think it would be better to have a single `char*` called e.g. `identity`? And it will be populated by whichever we happen to be using in this case?

Reply:
I think we should still use both `spiffe_id` and `hostname` (one could be `nullptr` if it is not populated) since using a single `identity`, we may lose the information about whether it is one or the other. HDYT?

Reply:
SGTM. (This is how `s2a_tsi_handshaker_result_create` is currently implemented: it sets `spiffe_id` and/or `hostname` to nullptr if they are not populated by the `SessionResp` message from the S2A service.)